src/lib.rs

```diff
@@ -198,138 +238,105 @@ impl<T> Stream for WebSocketStream<T>
 where
     T: AsyncRead + AsyncWrite,
 {
-    type Item = Message;
-    type Error = WsError;
-
-    fn poll(&mut self) -> Poll<Option<Message>, WsError> {
-        self.inner
+    type Item = Result<Message, WsError>;
+
+    fn poll_next(
+        mut self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+    ) -> Poll<Option<Self::Item>> {
+        let this = self.project();
+        this.inner
             .read_message()
             .map(Some)
             .to_async()
             .or_else(|err| match err {
-                WsError::ConnectionClosed => Ok(Async::Ready(None)),
-                err => Err(err),
+                WsError::ConnectionClosed => Poll::Ready(None),
+                err => Poll::Ready(Some(Err(err))),
             })
     }
 }
 
-impl<T> Sink for WebSocketStream<T>
+impl<T> Sink<Message> for WebSocketStream<T>
 where
     T: AsyncRead + AsyncWrite,
 {
-    type SinkItem = Message;
-    type SinkError = WsError;
+    type Error = WsError;
 
-    fn start_send(&mut self, item: Message) -> StartSend<Message, WsError> {
-        self.inner.write_message(item).to_start_send()
+    fn poll_ready(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        Poll::Ready(Ok(()))
     }
-
-    fn poll_complete(&mut self) -> Poll<(), WsError> {
-        self.inner.write_pending().to_async()
+    fn start_send(mut self: Pin<&mut Self>, item: Message) -> Result<(), Self::Error> {
+        self.inner.write_message(item)
     }
-
-    fn close(&mut self) -> Poll<(), WsError> {
-        self.inner.close(None).to_async()
+    fn poll_flush(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        match self.with_context(cx, |s| cvt(s.write_pending())) {
+            Ok(()) => Poll::Ready(Ok(())),
+            Err(ref e) if e.kind() == std::io::ErrorKind::WouldBlock => Poll::Pending,
+            Err(e) => Poll::Ready(Err(e)),
+        }
     }
-}
-
-/// Future returned from client_async() which will resolve
-/// once the connection handshake has finished.
-pub struct ConnectAsync<S: AsyncRead + AsyncWrite> {
-    inner: MidHandshake<ClientHandshake<S>>,
-}
-
-impl<S: AsyncRead + AsyncWrite> Future for ConnectAsync<S> {
-    type Item = (WebSocketStream<S>, Response);
-    type Error = WsError;
-
-    fn poll(&mut self) -> Poll<Self::Item, WsError> {
-        match self.inner.poll()? {
-            Async::NotReady => Ok(Async::NotReady),
-            Async::Ready((ws, resp)) => Ok(Async::Ready((WebSocketStream::new(ws), resp))),
+    fn poll_close(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        match self.with_context(cx, |s| cvt(s.close(None))) {
+            Ok(()) => Poll::Ready(Ok(())),
+            Err(ref e) if e.kind() == std::io::ErrorKind::WouldBlock => Poll::Pending,
```

Comment:
How does the task get woken up here if we get wouldblock?

Reply:
Not entirely sure on that. As mentioned above, taken from tokio-tls, so would have to ask there.

Reply:
Ok, yeah, I think this is wrong then. If you return Pending without an inner future returning Pending, you have to wake up the task somewhere. I don't know if there is a recommended way of doing that, but I usually store the Waker in an option and when other code in my object detects the right condition, it wakes up that task. In this case I doubt (haven't dug into it completely and it's not documented :( ) that tungstenite returns wouldblock here at all...

Reply:
Will look at tungstenite some more. If it never returns wouldblock, then we can nuke compat, and use tokio blocking around the read/write.

Reply:
Yeah, looks like any would blocks are converted into other types, which we can chain futures off of.

Reply:
I had the impression it won't return wouldblock for `close`. In principle it might for other things, but since we are working on AsyncRead/AsyncWrite, and we are have the impl of `std::Read`/`std::Write`, we could intercept the Pendings there and tungstenite would not have to return wouldblock.

One exception would be the `MessageQueueFull` error probably.